doc: documentation for ALTER SINK (#27594)
MaterializeInc/materialize: doc/user/content/sql/alter-sink.md

item that happened before the cutover timestamp and all the updates of the new
item that happened after the cutover timestamp. Care must be taken to ensure
that the old and the new items make sense when stitched together. See the
anomalies section for more details on what can be observed.

**Note**: In order to select a consistent timestamp Materialize waits for the
previous definition of the sink to make enough progress such that there is
overlap between the available timestamps of the desired new item. Attempting to
`ALTER` an unhealthy sink that can't make progress will result to the command
timing out.

### Valid schema changes

Materialize allows altering a sink to a new item provided that the new item
leads to a valid sink definition with the same rules as the `CREATE SINK`
statement.

When using the Avro format with a schema registry care must be taken such that
the generated Avro schema of the new item is compatible with the previously
published schema. When that is not the case the `ALTER SINK` statement will
succeed but the subsequent execution of the sink will result in errors and will
be unable to make progress.

> ### Possible anomalies

Reviewer: These are all standard operating procedure for a stream processor. I wonder if we want to call that out. We're "only" providing the same level of correctness as a sp here as opposed to our standard higher level of correctness.

Depending on the contents and state of the old and new items a few different
anomalies can be observed in the output topic as a result of altering a sink.
These anomalies occur because `ALTER SINK` causes Materialize to emit updates
from the new item only if they happen after the cutover timestamp.

**Anomaly 1: Topic contains stale value for a key**

As a consequence of not re-emitting a snapshot after `ALTER SINK`, all keys of
the collection will appear to have the value they had in the old collection
until an update happens to them. At that point the current value will be
published to the topic.

It is important that readers of the topic are prepared to handle an old value
for a key, for example by filling in additional columns with default values.

> Alternatively, forcing an update to all the keys after `ALTER SINK` will force

Reviewer 2: Does it make sense to mention that dropping and recreating the sink emits a complete snapshot to guild folks who should NOT use `ALTER SINK in` the right direction?

Reviewer 3: Is there a use case for `ALTER SINK` outside blue/green deployments right now? I'd like to frame this in the clearest possible way. If there isn't, no one should be manually running this command, and these docs should point very early on to the dbt workflow. I'm pushing some updates soon.

---
title: "ALTER SINK"
description: "`ALTER SINK` allows cutting a sink over to a new upstream relation without causing disruption to downstream consumers."
menu:
  main:
    parent: 'commands'
---

`ALTER SINK` allows cutting a sink over to a new upstream relation without
causing disruption to downstream consumers. This is useful in the context
of [blue/green deployments](/manage/dbt/development-workflows/#bluegreen-deployments).

## Syntax

{{< diagram "alter-sink.svg" >}}

## Details

> To alter the upstream relation a sink depends on and ensure continuity in data

Reviewer 4: ```suggestion
To alter a sink's upstream relation and ensure continuity in data
```

(I think "upstream" and "depends on" are redundant here.)

Reviewer 4: (Also @morsapaes the `CREATE SINK` documentation says "underlying relation" where we say "upstream relation" here—we should bring them in sync! I think I like the new "upstream relation" term better.)

Reviewer 4: Actually I feel like something about the wording is off here! It's not about altering _and_ ensuring continuity, it's about ensuring continuity _while_ altering, right?

```suggestion
To ensure continuity in data processing when altering a sink's upstream relation, Materialize must pick ...
```

processing, Materialize must pick a consistent cutover timestamp. When you
execute an `ALTER SINK` command, the resulting output will contain all the
updates that happened before the cutover timestamp for the old relation, as
well as all the updates that happened after the cutover timestamp for the new
relation.

{{< note >}}

> To select a consistent timestamp, Materialize waits for the previous definition
> of the sink to make enough progress such that there is overlap between the
> available timestamps of the newly specified upstream relation. Attempting to

Reviewer 4: The wording of the first sentence here is a little wonky. Tinkering a bit:

```suggestion
To select a consistent timestamp, Materialize must wait for the previous definition
of the sink to emit results up until the oldest timestamp at which the contents of
the new upstream relation are known. Attempting to
```

`ALTER` an unhealthy sink that can't make progress will result in the command
timing out.
{{</ note >}}

### Valid schema changes

For `ALTER SINK` to be successful, the newly specified relation must lead to a
valid sink definition with the same conditions as the original `CREATE SINK`
statement.

When using the Avro format with a schema registry, the generated Avro
schema for the new relation must be compatible with the previously published
schema. If that's not the case, the `ALTER SINK` command will succeed, but the
subsequent execution of the sink will result in errors and will not be able to

> make progress.

Reviewer 4: Maybe add a recommendation here: monitor your sink status via <this SQL query> or in the console after an alteration to be ensure it remains healthy.
